mffpy/header_block.py

```diff
@@ -75,86 +61,131 @@
-    fp.write(arr.tobytes())
-    pad_byte_len = hdr.header_size - 4 * (4 + 2 * hdr.num_channels)
-    # Pad either with zeros or with the magic `PADDING`
-    padding = PADDING if pad_byte_len == len(PADDING) \
-        else np.zeros(pad_byte_len, dtype=np.uint8).tobytes()
-    fp.write(padding)
+def read(fp: FileLike, format_str: str):
+    num_bytes = struct.calcsize(format_str)
+    byts = fp.read(num_bytes)
+    ans = struct.unpack(format_str, byts)
+    return ans if len(ans) > 1 else ans[0]
+
+
+def skip(fp: FileLike, n: int):
+    fp.seek(n, SEEK_CUR)
+
+
+def write(fp: FileLike, format_str: str, items: tuple):
+    pack = struct.pack(format_str, *items)
+    fp.write(pack)
+
+
+class HeaderBlock(_HeaderBlock):
+
+    HEADER_PRESENT = 1
+
+    def __new__(cls,
+                block_size: int,
+                num_channels: int,
+                num_samples: int,
+                sampling_rate: int,
+                header_size: Optional[int] = None):
+        """create new HeaderBlock instance
+
+        Parameters
+        ----------
+        block_size : byte size of the block
+        num_channels : channel count in the block
+        num_samples : sample count per channel in the block
+        sampling_rate : sampling_rate per channel in the block
+        header_size : byte size of the header (computed if None)
+        """
+        header_size = header_size or cls.compute_byte_size(num_channels)
+        return super().__new__(cls, header_size, block_size, num_channels,
+                               num_samples, sampling_rate)
+
+    @classmethod
+    def from_file(cls, fp: FileLike):
+        """return HeaderBlock, read from fp"""
+
+        # Each block starts with a 4-byte-long header flag which is
+        # * `0`: there is no header
+        # * `1`: it follows a header
+        if read(fp, 'i') == 0:
+            return None
+        # Read general information
+        header_size, block_size, num_channels = read(fp, '3i')
+        # number of 4-byte samples per channel in the data block
+        num_samples = (block_size//num_channels) // 4
+        # Read channel-specific information
+        nc4 = 4 * num_channels
+        # Skip byte offsets
+        skip(fp, nc4)
+        # Sample rate/depth: Read one skip, over the rest
+        # We also check that depth is always 4-byte floats (32 bit)
+        sampling_rate, depth = cls.decode_rate_depth(read(fp, 'i'))
+        skip(fp, nc4 - 4)
+        assert depth == 32, f"""
+        Unable to read MFF with `depth != 32` [`depth={depth}`]"""
+        # Skip the mysterious signal offset 2 (typically 24 bytes)
+        padding_byte_size = header_size - 16 - 2 * nc4
+        skip(fp, padding_byte_size)
+        return cls(
+            block_size=block_size,
+            header_size=header_size,
+            num_samples=num_samples,
+            num_channels=num_channels,
+            sampling_rate=sampling_rate,
+        )
+
+    def to_file(self, fp: FileLike):
```

Comment:
I understand naming this `to_file` for parity with `from_file`, but I'm wondering if we could have a better name here to indicate that the class is actually being written out to the file. Just my opinion though.

Reply:
So I checked the rest of the API in `mffpy`, and "xml_files" are also read by `.from_file()`.  However, writing seems to be done using `.write()` so I'm changing this accordingly.  We could change that in the future though if we come up with better verbs, i.e. `read_from()`, `.write_to()`.   What do you think?